docs/contracts.rst

```diff
@@ -172,6 +172,35 @@ Each Contract Factory exposes the following methods.
         >>> my_contract.estimateGas().multiply7(3)
         42650
 
+.. py:method:: Contract.prepareTransaction(transaction).myMethod(*args, **kwargs)
+
+    Prepare a transaction dictionary based on the contract function call specified. 
+
+    This method behaves the same as the :py:method::`Contract.transact` method,
+    with transaction details being passed into the first portion of the
+    function call, and function arguments being passed into the second portion.
+
+    .. note::
+        If ``web3.eth.defaultAccount`` is not set, this method will need another way
```

Comment:
I think it it's ok to return a transaction dict without a `nonce`.  In fact, I'm inclined to have `nonce` inclusion be opt-in, here's my reasoning.

* If I'm preparing a *generic* transaction that I want to send to many different contracts, I don't want `nonce` included.
* If I'm preparing a transaction to be sent once, not including nonce won't hurt because the `web3.eth.sendTransaction` API (and underlying `eth_sendTransaction` JSON-RPC) automatically add the proper nonce.

I think that fully excluding `nonce` from the return dict unless it's included in the `transaction` parameter would be just as easy to work with and would make using this API with different use cases easier.

Reply:
> i'm a little concerned how this would look for ...  prepare a single transaction, sign it and send it raw

This is a good instinct. We want to make sure the simple things stay easy. On the other hand, we risk confusing users who don't think about nonces explicitly.

Problems
---------

If someone prepares two transactions for an account, the second will clobber the first transaction, or they will end up with a confusing error about gas price not being high enough. For example:

```
transaction_dict1 = myContract.functions.myMethod().asdict()
transaction_dict2 = myContract.functions.myMethod().asdict()

signed = w3.eth.account.signTransaction(transaction_dict1, private_key)
w3.eth.sendRawTransaction(signed.rawTransaction)

signed = w3.eth.account.signTransaction(transaction_dict2, private_key)
w3.eth.sendRawTransaction(signed.rawTransaction)
# ^ error or overwrite
```

Note that with the current implementation, this scenario even breaks for the client-signed approach, like:

```
transaction_dict1 = myContract.functions.myMethod().asdict()
transaction_dict2 = myContract.functions.myMethod().asdict()

w3.eth.sendTransaction(transaction_dict1)
w3.eth.sendTransaction(transaction_dict2)
# ^ error or overwrite
```

A Weak Solution
---------

One solution here is to add another public "pre-signing" method to always be called immediately before offline signing. Something like:

```
transaction_dict1 = myContract.functions.myMethod().asdict()
transaction_dict2 = myContract.functions.myMethod().asdict()

# toSignableTransaction must be called immediately before signing and sending,
# ... and called after the most recently sent transaction

signable_transaction = w3.eth.toSignableTransaction(transaction_dict1)
signed = w3.eth.account.signTransaction(signable_transaction, private_key)
w3.eth.sendRawTransaction(signed.rawTransaction)

signable_transaction = w3.eth.toSignableTransaction(transaction_dict2)
signed = w3.eth.account.signTransaction(signable_transaction, private_key)
w3.eth.sendRawTransaction(signed.rawTransaction)
```

It adds an extra step, but might save us some headaches with people who prepare transactions and don't sign/send them immediately.

`toSignableTransaction` still isn't trivial in this case, because `getTransactionCount` will not recognize the pending transactions afaik, so we'd have to be even more clever about tracking the pending transactions.

A Better Solution
----------------

Explicitly calling `getTransactionCount` is actually the best option I can see. That's because tracking nonces is so hard that we should not even try to do it automatically, we should ask the user to figure it out explicitly, like:

```
transaction_dict1 = myContract.functions.myMethod().asdict()
transaction_dict2 = myContract.functions.myMethod().asdict()

transaction_dict1['nonce'] = w3.eth.getTransactionCount(transaction_dict1['from'])
transaction_dict2['nonce'] = transaction_dict1['nonce'] + 1

signed = w3.eth.account.signTransaction(transaction_dict1, private_key)
w3.eth.sendRawTransaction(signed.rawTransaction)

signed = w3.eth.account.signTransaction(transaction_dict2, private_key)
w3.eth.sendRawTransaction(signed.rawTransaction)
```

Then this will work transparently again, without deleting the auto-generated nonce values:

```
transaction_dict1 = myContract.functions.myMethod().asdict()
transaction_dict2 = myContract.functions.myMethod().asdict()

w3.eth.sendTransaction(transaction_dict1)
w3.eth.sendTransaction(transaction_dict2)
```

A Stretch to get the Best of Both Worlds
-------------
Maybe we can still get a clean single solution with a special keyword, like:

```
transaction_dict = myContract.functions.myMethod().asdict({'nonce': 'current'})

signed = w3.eth.account.signTransaction(transaction_dict, private_key)
w3.eth.sendRawTransaction(signed.rawTransaction)
```

Hopefully `current` is enough of a visual signal to people that two consecutive calls will be broken. (I'm not convinced about this part, I'm going to let this marinate and come back to it)

Reply:
Ah yes, i hadn't really considered tracking nonces... I think you are definitely right in that we should not even attempt it. It will only introduce more magic that is hard to figure out... 

In your "Better Solution" nonce can also be specified inline as a shortcut:
```
transaction_dict1 = myContract.functions.prepareTransaction({
'nonce':  w3.eth.getTransactionCount(myAccount)
}).myMethod()
```
The good thing about this is that it's pretty clearly presentable in an example with an extra mention of how nonce management is left upto the user. 
I think we are all in agreement then that for now nonce should not be handled by the method. I will make that change.

Reply:
How about instead of special keyword magic we just allow an extra kwarg to that API:

```python
transaction_dict = myContract.functions.myMethod().asdict({'from': web3.eth.coinbase}, add_nonce=True)
```

Reply:
(duplicating comment here)

How about instead of special keyword magic we just allow an extra kwarg to that API:

```python
transaction_dict = myContract.functions.myMethod().asdict(
    {'from': web3.eth.coinbase},
    add_nonce=True,
)
```

Reply:
Happy to add that @pipermerriam, was the previous behaviour of how nonce was added suitable (only with this extra flag ofcourse). sorry for wiping that out pre-emptively :slightly_frowning_face: As reminder, it was
1. Nonce of provider default account
2. Nonce of account specified in call 'from' attribute
Also we need to then explain in docs that there will be potential nonce conflicts with this feature

Reply:
@monokh Rather than go back and forth on how to specify auto-nonce, let's just leave it out altogether. We can come back to it later in another issue, if it becomes important. You're example explicitly using `getTransactionCount` looks good.

Reply:
I'm pretty sure his comment was "if auto-nonce stays in, maybe this way would look better", rather than "I'm advocating for auto-nonce to go back in".

Reply:
Oh and that means the docs should change too. `from` or `nonce` no longer need to be specified here.

Reply:
Ok, let me know if you would like to revisit this one :) happy to help with this in another PR